libs/medic-users-meta/migrations/202105171933.do.86.usersMeta.sql

```diff
@@ -0,0 +1,59 @@
+-- Drop views with wrong indexes and definitions that were
+-- created at 202102191153.do.54-create-couchdb-users-meta-table.sql
+DROP MATERIALIZED VIEW IF EXISTS useview_feedback;
+DROP MATERIALIZED VIEW IF EXISTS useview_telemetry;
+
+-- The views are created again
+CREATE MATERIALIZED VIEW useview_feedback AS
+SELECT
+    doc->>'_id' AS uuid,
+    doc#>>'{meta,source}' AS SOURCE,
+    doc#>>'{meta,url}' AS url,
+    doc#>>'{meta,user,name}' AS user_name,
+    doc#>>'{meta,time}' AS period_start,
+    COALESCE(doc#>>'{info,cause}',doc->>'info') AS cause,
+    doc#>>'{info,message}' AS message
+FROM
+    couchdb_users_meta
+WHERE
+    doc->>'type'='feedback';
+
+CREATE UNIQUE INDEX idx_useview_feedback_uuid ON useview_feedback(uuid);  --> Only to allow the refresh of the view CONCURRENTLY
+CREATE INDEX idx_useview_feedback_period_start_user ON useview_feedback(period_start,user_name);
+
+CREATE MATERIALIZED VIEW useview_telemetry AS
+SELECT
+  doc->>'_id' AS uuid,
+  CONCAT_WS(                                --> Date concatenation from JSON fields, eg. 2021-5-17
+    '-',
+    doc#>>'{metadata,year}',                --> year
+    CASE                                    --> month of the year
+      WHEN
+        string_to_array(substring(doc#>>'{metadata,versions,app}' FROM '(\d+.\d+.\d+)'),'.')::int[] < '{3,8,0}'::int[]
```

Comment:
Converting semver to int for comparing won't always yield correct results, for example : `3.13.3` compared to `3.7.12`. 
Let's hope this won't bite us later on, hopefully we won't have > 10 patch releases for a single minor ever.

Reply:
:thinking: yes in fact I couldn't test this properly in my local environment, the field `metadata.versions.app` is set to `"unknown"` in CHT 3.10 and 3.11, either if the record comes from desktop or mobile (at least in my local tests), but for those cases the result ends up with months 1-indexed.

I'll add one more condition: now that in 3.12+ we will add the `metadata.day` field, at least for these cases the month will be always 1-indexed, otherwise the condition above will be used.

Reply:
Oh, didn't know that. The change to compute the month was introduced in 3.2.0 with errors and @derickl provided the patch because the condition was the opposite to the one needed as you also pointed out. I'm not sure what value to expect there , but I guess it is a string value that is cast by that `SELECT` clause. In the case you mention : `"app": "3.13.3"`, the result from the expression seems to be right. Here is the query with that value that I tested in a `psql` session:

```
postgres=# SELECT string_to_array(substring('3.13.3' FROM '(\d+.\d+.\d+)'),'.')::int[] < '{3,8,0}'::int[];
 ?column? 
----------
 f
(1 row)
```

I also tested it locally editing records in my CouchDB with the values `"3.13.3"` for one record, and another with `"3.7.1"`, and in both cases I got the result expected in Postgres after run the process. Also testing the `"unknown"` value and no value at, the results are the expected:

```
postgres=# SELECT CASE WHEN string_to_array(substring('unknown' FROM '(\d+.\d+.\d+)'),'.')::int[] < '{3,8,0}'::int[] THEN TRUE ELSE FALSE END;
 case 
------
 f
(1 row)
postgres=# SELECT CASE WHEN string_to_array(substring(NULL FROM '(\d+.\d+.\d+)'),'.')::int[] < '{3,8,0}'::int[] THEN TRUE ELSE FALSE END;
 case 
------
 f
(1 row)
```

So I'm not going to add that extra condition checking the day of the month after all.